### PR TITLE
pythonPackages.scs: init at 2.1.1

### PR DESCRIPTION
**Pull request by drewrisinger**
###### Motivation for this change

Add dependency for qiskit (cvxpy's dependency).
Python interface for C Splitting Cone Solver. C Package in tree at ``lib.science.math.scs``.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by drewrisinger** [comment]
@GrahamcOfBorg build python3Packages.scs

File: pkgs/development/python-modules/scs/default.nix [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  version = "2.1.1";
+
+  disabled = pythonOlder "3.5";
+
+  src = fetchFromGitHub {
+    owner = "bodono";
+    repo = "scs-python";
+    rev = "f02abdc0e2e0a5851464e30f6766ccdbb19d73f0"; # need to choose commit manually, untagged
+    sha256 = "01ghvyylxql7jvdcwy65wf7y5qykqyqkbysy402zfhgmm9dxizgv";
+    fetchSubmodules = true;
```
Comment (bhipple) [inline pkgs/development/python-modules/scs/default.nix]: Ihttps://github.com/bodono/scs-python/blob/master/setup.py
If you look at what the upstream src repo and `setup.py` are doing, it appears they're shipping the `scs` C library as a git submodule, then having python recompile it from scratch.

In the nix distro, we generally prefer to have the C package stand-alone, then have the python package just link the package as a proper distribution dependency without vendoring.  An example of a package that does this is pyarrow:
https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/python-modules/pyarrow/default.nix

**Review by bhipple — CHANGES_REQUESTED** [review]

**Comment by drewrisinger** [comment]
I had originally done that but wasn't sure what the best/most replicable
way of moving the C source to source/scs was. I will look at the quoted
example.

File: pkgs/development/python-modules/scs/default.nix (hunk 1 above)
Comment (drewrisinger) [reply pkgs/development/python-modules/scs/default.nix]: I'm trying to do this, and I'm stuck with getting the proper nix path. I have the following:
```
prePatch = ''
    cp -r ${scs.src} ./scs
'';
```
Issue is that the actual source is stored at something like ``${scs.src}/j4i4blqll2d3wrzzllh1b2hz52s2iybs-source``, based on the derivation name/hash, I believe. Any tips on how to get this path?

Note that this is different from ``pyarrow`` b/c that includes the python source in the same directory.

Comment (bhipple) [reply pkgs/development/python-modules/scs/default.nix]: `${scs.src}` should work; you can view it like this:
```
λ brh nixpkgs →  tree $(nix-build -A scs.src)
/nix/store/q1alada06jhjpk7cnqjsy92pdlzpyj6n-source
├── include
│   ├── accel.h
│   ├── cones.h
...
```

You probably have to rename this one to `python-scs` to avoid infinite recursion on the `scs` name.  
https://github.com/bhipple/nixpkgs/commit/494f2efd434e8d87623b93c7b73d61c9935d8b44

In an ideal world, we would rely on just `${scs}` (the already-compiled binary package with SO) in the python package, rather than the src directly, and re-use the SO's and headers from the installed `${scs}`.  Doing so may require more patching to the setup.py, but I'd give it a shot to see if it isn't too hard, as that's the "best" solution.  In lieu of that, vendoring the binaries but sharing the src is second best, and if THAT doesn't work then we're back to the initial PR ... with the knowledge that these things may not stay in sync :)

Comment (bhipple) [reply pkgs/development/python-modules/scs/default.nix]: When I do all of the above, I find:

```
src/scsmodule.c: In function ‘csolve’:
src/scsmodule.c:331:20: error: ‘ScsSettings’ {aka ‘struct SCS_SETTINGS’} has no member named ‘write_data_filename’
  331 |           &(d->stgs->write_data_filename),
      |                    ^~
```
which suggests that our version of `scs` is not in sync with what `python-scs` expects?

Comment (drewrisinger) [reply pkgs/development/python-modules/scs/default.nix]: Thanks for the help & examples. I'll commit the above and try to get the code/build working. ^scs version might be due to not having #78896 in your branch?

**Comment by drewrisinger** [comment]
@GrahamcOfBorg eval

**Comment by drewrisinger** [comment]
@bhipple suggested changes made. Please review/approve.

File: pkgs/top-level/python-packages.nix [hunk 2]
```diff
@@ -4980,6 +4980,8 @@ in {
   python-simple-hipchat = callPackage ../development/python-modules/python-simple-hipchat {};
   python_simple_hipchat = self.python-simple-hipchat;
 
+  python-scs = callPackage ../development/python-modules/python-scs { };
```
Comment (jonringer) [inline pkgs/top-level/python-packages.nix]: this should try to match what is published on pypi, if this is https://pypi.org/project/scs/ it should be
```suggestion
  scs = callPackage ../development/python-modules/python-scs { scs = pkgs.scs; };
```

>  One of the arguments is the name of the package, which consists of a basename (generally following the name on PyPi) and a version. 

from: https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#packaging-a-library

Comment (jonringer) [reply pkgs/top-level/python-packages.nix]: also, the directory name should change.

I know this kind of contradicts what @bhipple said, but the python-module name should align with upstream intentions.

Comment (jonringer) [reply pkgs/top-level/python-packages.nix]: if a package requires scs, then declaring it as a dependency in their package.nix should be very similar to their requirements.txt or setup.py

File: pkgs/development/libraries/science/math/scs/default.nix [hunk 3]
```diff
@@ -4,6 +4,7 @@ stdenv.mkDerivation rec {
   pname = "scs";
   version = "2.1.1";
 
+  # NOTE: check pythonPackages.python-scs when updating.
```
Comment (jonringer) [inline pkgs/development/libraries/science/math/scs/default.nix]: this is likely uneeded, most reviewers will do `nix-review` which will pick up that this derivation has changed.

at the very least it should go in a different commit

File: pkgs/top-level/python-packages.nix [hunk 4]
```diff
@@ -4980,6 +4980,8 @@ in {
   python-simple-hipchat = callPackage ../development/python-modules/python-simple-hipchat {};
   python_simple_hipchat = self.python-simple-hipchat;
 
+  scs = callPackage ../development/python-modules/scs { };
```
Comment (jonringer) [inline pkgs/top-level/python-packages.nix]: to avoid infinite recursion
```suggestion
  scs = callPackage ../development/python-modules/scs { scs = pkgs.scs; };
```

Comment (drewrisinger) [reply pkgs/top-level/python-packages.nix]: True, thanks.

**Comment by jonringer** [comment]
@GrahamcOfBorg build pythonPackages.scs
@GrahamcOfBorg build python3Packages.scs
@GrahamcOfBorg build python38Packages.scs

**Review by bhipple — APPROVED** [review]
Looks much nicer now!  Thanks for the help with best practices too, @jonringer.
